bundles/org.openhab.binding.airgradient/src/main/java/org/openhab/binding/airgradient/internal/handler/AirGradientAPIHandler.java

```diff
@@ -0,0 +1,401 @@
+                measure.tvoc = metric.getValue();
+            } else if (metric.getMetricName().equals("airgradient_tvoc_raw_index")) {
+                measure.tvocIndex = metric.getValue();
+            } else if (metric.getMetricName().equals("airgradient_nox_index")) {
+                measure.noxIndex = metric.getValue();
+            } else if (metric.getMetricName().equals("airgradient_temperature_degc")) {
+                measure.atmp = metric.getValue();
+            } else if (metric.getMetricName().equals("airgradient_humidity_percent")) {
+                measure.rhum = metric.getValue();
+            }
+
+            if (metric.getLabels().containsKey("id")) {
+                String id = metric.getLabels().get("id");
+                measure.serialno = id;
+                measure.locationId = id;
+                measure.locationName = id;
+            }
+
+            if (metric.getLabels().containsKey("airgradient_serial_number")) {
+                String id = metric.getLabels().get("airgradient_serial_number");
+                measure.serialno = id;
+                measure.locationId = id;
+                measure.locationName = id;
+            }
+        }
+
+        return Arrays.asList(measure);
+    }
+
+    private List<Measure> parseJson(String stringResponse) {
+        List<@Nullable Measure> measures = null;
+        if (stringResponse.startsWith("[")) {
+            // Array of measures, like returned from the AirGradients API
+            Type measuresType = new TypeToken<List<@Nullable Measure>>() {
+            }.getType();
+            measures = gson.fromJson(stringResponse, measuresType);
+        } else if (stringResponse.startsWith("{")) {
+            // Single measure e.g. if you read directly from the device
+            Type measureType = new TypeToken<Measure>() {
+            }.getType();
+            Measure measure = gson.fromJson(stringResponse, measureType);
+            measures = new ArrayList<@Nullable Measure>(1);
+            measures.add(measure);
+        }
+
+        if (measures != null) {
+            List<@Nullable Measure> nullableMeasuresWithoutNulls = measures.stream().filter(Objects::nonNull).toList();
+            List<Measure> measuresWithoutNulls = new ArrayList<Measure>(nullableMeasuresWithoutNulls.size());
+            for (@Nullable
+            Measure m : nullableMeasuresWithoutNulls) {
+                if (m != null) {
+                    measuresWithoutNulls.add(m);
+                }
+            }
+
+            return measuresWithoutNulls;
+        }
+
+        return Collections.emptyList();
+    }
+
+    @Override
+    public void dispose() {
+        ScheduledFuture<?> p = pollingJob;
+        if (p != null) {
+            p.cancel(true);
+            pollingJob = null;
+        }
+    }
+
+    private @Nullable String generatePingUrl() {
+        AirGradientAPIConfiguration config = getConfiguration();
```

Comment:
Better to store the configuration on class level at init. Whenever the config is changed, the framework causes the binding to re-init. When declared it can be assigned a default value, then it is non-null.

Reply:
Sotred